### PR TITLE
locksmithctl, lock: implement groups

### PR DESCRIPTION
**Pull request by mischief**

### CODE REVIEW

**Comment by mischief** [comment]
cc @philips

**Comment by mischief** [comment]
fixes #70

File: lock/etcd.go [hunk 1]
```diff
@@ -33,11 +34,17 @@ const (
...
+	client  etcd.EtcdClient
+	keypath string
 }
 
-func NewEtcdLockClient(ec etcd.EtcdClient) (client *EtcdLockClient, err error) {
-	client = &EtcdLockClient{ec}
+func NewEtcdLockClient(ec etcd.EtcdClient, group string) (client *EtcdLockClient, err error) {
+	key := SemaphorePrefix
+	if group != "" {
+		key = path.Join(keyPrefix, "groups", group, "semaphore")
```
Comment (crawford) [inline lock/etcd.go]: Can you pull these strings into constants like SemaphorePrefix?

**Comment by mischief** [comment]
@crawford PTAL. also escaped the group string so / or other characters can be used in it.

File: lock/etcd.go [hunk 2]
```diff
@@ -27,17 +29,25 @@ const (
...
 // structs through etcd.
 type EtcdLockClient struct {
-	client etcd.EtcdClient
+	client  etcd.EtcdClient
+	keypath string
 }
 
-func NewEtcdLockClient(ec etcd.EtcdClient) (client *EtcdLockClient, err error) {
-	client = &EtcdLockClient{ec}
+func NewEtcdLockClient(ec etcd.EtcdClient, group string) (client *EtcdLockClient, err error) {
```
Comment (jonboulle) [inline lock/etcd.go]: Can you add a docstring detailing the use of group?

File: lock/etcd.go [hunk 3]
```diff
@@ -27,17 +29,25 @@ const (
 	keyPrefix       = "coreos.com/updateengine/rebootlock"
 	holdersPrefix   = keyPrefix + "/holders"
```
Comment (jonboulle) [inline lock/etcd.go]: can you kill this? appears to be unused

**Comment by jonboulle** [comment]
Seems reasonable, some test cases would be nice

**Comment by marineam** [comment]
Could you update the README? I don't actually know what this is supposed to do.

File: lock/etcd.go [hunk 4]
```diff
@@ -25,19 +27,29 @@ import (
 
 const (
 	keyPrefix       = "coreos.com/updateengine/rebootlock"
-	holdersPrefix   = keyPrefix + "/holders"
 	SemaphorePrefix = keyPrefix + "/semaphore"
+	groupPath       = "groups"
+	semaphorePath   = "semaphore"
```
Comment (vcaputo) [inline lock/etcd.go]: These don't seem appropriately named to me, they're path components, but a single component of a path isn't generally referred to as a path.

Perhaps s/Path/Branch/g?

And make SemaphorePrefix = keyPrefix + semaphoreBranch?

File: README.md [hunk 5]
```diff
@@ -80,6 +80,29 @@ Old: 1
...
+`locksmithd` coordinates the reboot lock in groups of machines. The default
+group is "", or the empty string. `locksmithd` will only coordinate the reboot
+lock with other machines in the same group.
+
+The purpose of groups is to allow faster updating of certain sets of machines
+while maintaining availability of certain services. For example, in a cluster
+of 5 CoreOS machines, if you have 2 load balancers and run `locksmithctl set-max 2`,
+then it is possible that both load balancers would be rebooted at the same time,
+interrupting the service they provide. However, if the load balancers are put
+into their own group named "lb", the other 3 machines will remain in the default
```
Comment (marineam) [inline README.md]: Isn't clear: is set-max 2 per group or cluster-wide? If per group and you did set-max 2 in the default group would it then be possible for three hosts to reboot at the same time?

**Comment by marineam** [comment]
So far looks good but I agree with the needs tests suggestion.

**Comment by mischief** [comment]
@marineam did you review the ones i added? should i add more test cases for get/set?

**Comment by marineam** [comment]
Oh, I just missed that the changes in the test code was indeed using
non-default groups. That should be fine then.

So LGTM but would like confirmation from crawford or jon

File: lock/etcd.go [hunk 6]
```diff
@@ -25,19 +27,29 @@ import (
 
 const (
 	keyPrefix       = "coreos.com/updateengine/rebootlock"
-	holdersPrefix   = keyPrefix + "/holders"
-	SemaphorePrefix = keyPrefix + "/semaphore"
+	groupBranch     = "groups"
+	semaphoreBranch = "semaphore"
+	SemaphorePrefix = keyPrefix + "/" + semaphoreBranch
```
Comment (jonboulle) [inline lock/etcd.go]: on actually thinking about this SemaphorePrefix is just a bad name right? what about defaultSemaphorePath?

Comment (jonboulle) [inline lock/etcd.go]: or defaultSemaphoreKey

**Comment by jonboulle** [comment]
Something I didn't really think about is the experience of running `locksmithctl` on systems where locksmithd is configured to point to a group - seems like an easy way for users to shoot themselves in the foot if they forget to use the argument. What about having locksmithctl inspect the conf and use what it finds as the default?

**Comment by mischief** [comment]
@jonboulle yes. this is a definite potential pitfal. but, what is "the conf"? afaict, locksmithd is exclusively configured through systemd unit dropins that set environment variables. perhaps LOCKSMITHCTL_GROUP can be set in /etc/environment? :-)

**Comment by jonboulle** [comment]
@mischief uhh, yeah, what you said ;-)

**Comment by marineam** [comment]
Or use update.conf, isn't that already used for strategy?

**Comment by mischief** [comment]
Yes. The systemd unit uses update.conf. however the key group is already
used for updateengine and LOCKSMITHD_GROUP may be confusing...

**Comment by marineam** [comment]
Sounds then like locksmith's group concept needs a different name, not a
new config file. Configuring locksmith in even more locations sounds worse.

**Comment by crawford** [comment]
I just talked this over with @mischief and I'm not convinced the user will easily shoot themselves in the foot. The only case I could come up with is the user setting the max-rebooters option on the default group instead of the current machine's group. I doubt users are going to log into a load balancer to configure locksmith though. They will probably use an edge node and they should be well aware of which machines are in what group (they would have to manually set this up for starters).

**Comment by marineam** [comment]
@crawford are you suggesting that locksmithctl doesn't need to use a config file to ensure it uses the same group as locksmithd? I personally know I would forget about passing group to `locksmithctl status`, unless of course locksmithctl status always reports on all groups unless told to look at only one.

Another way to deal with keeping locksmithd and locksmithctl in sync: locksmithctl becomes a dbus clinet and locksmithd does everything. That is a much bigger project though.

**Comment by marineam** [comment]
locksmithctl being a dbus client would be a huge win for etcd setups with auth and/or client certs.

**Comment by crawford** [comment]
@marineam yes. I don't think locksmithctl needs the config. Maybe if it reported all of the group statuses by default it would be much more obvious to the user.

Setting up the client to talk to the daemon instead of etcd directly also crossed my mind. What if the user isn't running the daemon though? They could be running it from a non-coreos system or one with reboots disabled.

**Comment by marineam** [comment]
@crawford I think in order to support secure etcd setups we are going to need to share configuration between locksmithd and locksmithctl and dbus would make it possible for core and other users to query status while keeping the etcd client cert/key only accessible by root.

I think we can move forward with this change as-is but should look into dbus in the future. Yes it means locksmithctl is no longer a stand-alone thing but does anyone actually need it to be?

**Comment by mischief** [comment]
@marineam LGTY?

**Comment by marineam** [comment]
Yeah, LGTM, we can figure out the rest later.

**Pull request merged**